fem/gridfunc.cpp

```diff
@@ -1467,15 +1616,63 @@ void GridFunction::GetGradients(ElementTransformation &tr,
 }
 
 void GridFunction::GetVectorGradient(
-   ElementTransformation &tr, DenseMatrix &grad) const
+   ElementTransformation &T, DenseMatrix &grad) const
 {
-   MFEM_ASSERT(fes->GetFE(tr.ElementNo)->GetMapType() == FiniteElement::VALUE,
-               "invalid FE map type");
-   DenseMatrix grad_hat;
-   GetVectorGradientHat(tr, grad_hat);
-   const DenseMatrix &Jinv = tr.InverseJacobian();
-   grad.SetSize(grad_hat.Height(), Jinv.Width());
-   Mult(grad_hat, Jinv, grad);
+   switch (T.ElementType)
+   {
+      case ElementTransformation::ELEMENT:
+      {
+         MFEM_ASSERT(fes->GetFE(T.ElementNo)->GetMapType() ==
+                     FiniteElement::VALUE, "invalid FE map type");
+         DenseMatrix grad_hat;
+         GetVectorGradientHat(T, grad_hat);
+         const DenseMatrix &Jinv = T.InverseJacobian();
+         grad.SetSize(grad_hat.Height(), Jinv.Width());
+         Mult(grad_hat, Jinv, grad);
+      }
+      break;
+      case ElementTransformation::BDR_ELEMENT:
+      {
+         // In order to capture the normal component of the gradient we
+         // must evaluate it in the neighboring element.
+         FaceElementTransformations * FET =
+            fes->GetMesh()->GetBdrFaceTransformations(T.ElementNo);
+
+         // Boundary elements and Boundary Faces may have different
+         // orientations so adjust the integration point if necessary.
+         int o = 0;
+         if (fes->GetMesh()->Dimension() == 3)
+         {
+            int f;
+            fes->GetMesh()->GetBdrElementFace(T.ElementNo, &f, &o);
+         }
+
+         IntegrationPoint fip;
+         be_to_bfe(FET->GetGeometryType(), o, T.GetIntPoint(), fip);
+
+         FET->SetIntPoint(&fip);
+         ElementTransformation & T1 = FET->GetElement1Transformation();
+
+         GetVectorGradient(T1, grad);
```

Comment:
The integration point is never transferred from the face (i.e. from `fip`) to the volume. This issue seems to also affect the other methods that are extended in this branch.

Reply:
The new implementation of `FaceElementTransformations::SetIntPoint` transforms and sets the integration points in the volume transformations as well.

Reply:
Okay, I didn't realize there's now a new version of `SetIntPoint` in `FaceElementTransformations`. This is then not an issue.

Reply:
One issue with this new version of `FaceElementTransformations::SetIntPoint` is that now most of the DG integrators will repeat these transformations -- they do these transformations explicitly, if needed.

Reply:
One simple solution is to rename the new `FaceElementTransformations::SetIntPoint` to something like `FaceElementTransformations::SetAllIntPoints`. That way the DG integrators will not need to change and it will be more clear that we set all (face+element) quadrature point with the new name. What do you think @mlstowell?

Reply:
The DG integrators should not need to change.  They access `FET->Face->SetIntPoint` which does not compute and set the volume integration points.  I wouldn't mind adding a boolean argument to `FaceElementTransformations::SetIntPoint(const IntegrationPoint *ip, bool transform_all = true)` to disable this feature.

Reply:
You are right, in some places the deprecated `Face` member is used -- these should probably be updated. However, in many places `FaceElementTransformations::SetIntPoint` is used -- I see 5 places in `bilininteg.cpp`, 4 in `lininteg.cpp` and a few more.

Adding a default parameter does not help when reading the code -- when I see a call to `SetIntPoint` with one parameter I'll have to remember to check the type of the object to understand what is actually being called.

Reply:
You make a valid point, It will take some getting used to.  From my point of view this is a natural consequence of promoting `FaceElementTransformations` from a struct to a class.  When it was a struct it was natural to configure the various pieces separately because they were separate objects which just happened to be passed around as a struct for convenience.  Now that they are parts of a single object it makes sense, at least to me, to configure the object as a whole by default.  Adding more fine-grained control for experts, as we do with the configuration mask, is the right choice, I think.  But it does require more care.

However, I agree that readability is very important.  How about this:
```
/// Sets the face integration point
void SetFaceIntPoint(const IntegrationPoint *ip);

/// Sets the integration point in element 1
void SetElement1IntPoint(const IntegrationPoint *ip);

/// Sets the integration point in element 2
void SetElement2IntPoint(const IntegrationPoint *ip);

/// Set the integration point in the face and elements 1 and 2 if present
void SetIntPoint(const IntegrationPoint *ip);
```

Reply:
I don't mind adding the first three methods, even though they may not be needed at the moment.

However, for the last method, I think it is best to use a different name and keep `SetIntPoint` as inherited from the base class `ElementTransformation`.

Reply:
I'm sorry, @v-dobrev, but I have to disagree.  We must think of the consequences of using these methods incorrectly.  In the current implementation if a developer calls `SetIntPoint` when they don't actually need the element transforms configured then they take a performance hit.  This is not ideal.  If we only set the point in the `Face` object but the developer expects the entire object to be usable there are two possible outcomes.  First, the `ElementTransformation` might crash when it accesses the NULL `IntegrationPoint` pointer.  This is ok but not ideal.  The more likely outcome is that the `ElementTransformation` has been used before and has a non-NULL `IntegrationPoint` pointer.  In this case the `ElementTransformation` will quietly return the wrong results. This type of error is a nightmare to track down if it is even noticed (and that's a BIG "if").

Reply:
@mlstowell, by using the same name, you are trying to change the meaning of the method `SetIntPoint` in a derived class and that function is (intentionally, by design) not virtual. Derived classes are not supposed to do that because it creates confusion, as I mentioned above.

If you want to have a "safe" method, you can add a new virtual method (with a different name) in the base class and override that in derived classes. This is also a "safer" solution because it will work correctly even if you call the method by a reference/pointer to the base class.

In any case, I think we your input, @psocratis and @vladotomov, to help resolve this disagreement.

Reply:
What is the reason for `SetIntPoint` not being virtual?

Reply:
@mlstowell, I can give you my reasons but at this point it is not going to lead us anywhere.

I think we should wait for the reviewers' (@psocratis and @vladotomov) input.

Reply:
That's fine.  I'm not trying to be argumentative.  I was just curious.

Reply:
I am kind of torn between the two solutions... 
May I ask the following?

1) How bad is the performance hit in the case of the current implementation?
2) Is there a significant performance hit under what @mlstowell suggests for DG integrators? 
3) Finally, is there a real downside of using a new FaceElementTransformations::(Some different name other than SetIntPoint) ? 

If the answer to (3) is no, I am leaning towards that. 

Socratis